### PR TITLE
 Error on opening a split zim file on external storage.

### PR DESCRIPTION
**Pull request by MohitMaliFtechiz**
Issue #3605

 * Added test cases for testing whether the ZimFileReader can open the splitted zim file or not.
 * Currently the issue has not been fixed, we will mark PR ready for review when the issue is fixed on upstream.

### CODE REVIEW

**Comment by kelson42** [comment]
@mgautierfr will release new version of `libzim` and `java-libkiwix` in the next hours.
@MohitMaliDeveloper Be ready to use and put this PR in review.

**Comment by kelson42** [comment]
@MohitMaliFtechiz I have implemented libzim 9.2.1 in [java-libkiwix](https://github.com/kiwix/java-libkiwix/pull/94). Does that solve the problem?

**Comment by MohitMaliFtechiz** [comment]
@kelson42 Thanks for integrating the libzim `9.2.1` in the `java-libkiwix`. I have tried to load the splitted zim files with the latest bindings on `java-libkiwix` and now splitted zim files are loading. However, I have encountered an error in Android with zim-tools' splitted zim file, it was throwing the `Corrupted block detected` error when getting the `articleCount` and `mediaCount` of this splitted zim file, and due to this it was not showing in the `LibraryScreen`. I have placed a fix for this on the Android side, and now it starts showing the zim file in the library screen but without the media and article count.

Zim chunks:
1. wikipedia_ar_all_nopic_2023-11.zimaa (2.1 GB (2,14,74,74,503 bytes)
2. wikipedia_ar_all_nopic_2023-11.zimab (2.1 GB (2,14,74,71,095 bytes)
3. wikipedia_ar_all_nopic_2023-11.zimac (1.3 GB (1,28,14,26,182 bytes)
4. wikipedia_ar_all_nopic_2023-11.zimad (3.1 GB (3,14,05,16,010 bytes)
5. wikipedia_ar_all_nopic_2023-11.zimae (0 bytes)

This error was due to there is a `wikipedia_ar_all_nopic_2023-11.zimae` part with 0 bytes. Because of this, libzim was throwing a `Corrupted block detected` error. However, we can read the zim file content. I have tried to delete this part from the device and after that, libzim starts returning the `mediaCount` and articleCount` for this splitted zim file.


| With 0 bytes file  | Without 0 bytes file |
| ------------- | ------------- |
| BeforeDeleting0bytesFile  | AfterDeleting0BytesZimFilePart  |


```kotlin
java.lang.Exception: Corrupted block detected
2024-05-03 12:11:37.364 14087-14279 System.err              pid-14087                            W  	at org.kiwix.libzim.Archive.getArticleCount(Native Method)
2024-05-03 12:11:37.365 14087-14279 System.err              pid-14087                            W  	at org.kiwix.kiwixmobile.core.reader.ZimFileReader.getArticleCount(ZimFileReader.kt:169)
2024-05-03 12:11:37.365 14087-14279 System.err              pid-14087                            W  	at org.kiwix.kiwixmobile.core.reader.ZimFileReader.toBook(ZimFileReader.kt:324)
2024-05-03 12:11:37.365 14087-14279 System.err              pid-14087                            W  	at org.kiwix.kiwixmobile.core.zim_manager.fileselect_view.adapter.BooksOnDiskListItem$BookOnDisk.<init>(BooksOnDiskListItem.kt:67)
2024-05-03 12:11:37.365 14087-14279 System.err              pid-14087                            W  	at org.kiwix.kiwixmobile.core.StorageObserver$convertToBookOnDisk$1.invokeSuspend(StorageObserver.kt:59)
2024-05-03 12:11:37.365 14087-14279 System.err              pid-14087                            W  	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
2024-05-03 12:11:37.365 14087-14279 System.err              pid-14087                            W  	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
2024-05-03 12:11:37.366 14087-14279 System.err              pid-14087                            W  	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:279)
2024-05-03 12:11:37.366 14087-14279 System.err              pid-14087                            W  	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85)
2024-05-03 12:11:37.367 14087-14279 System.err              pid-14087                            W  	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
2024-05-03 12:11:37.367 14087-14279 System.err              pid-14087                            W  	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source:1)
2024-05-03 12:11:37.367 14087-14279 System.err              pid-14087                            W  	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
2024-05-03 12:11:37.367 14087-14279 System.err              pid-14087                            W  	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source:1)
2024-05-03 12:11:37.367 14087-14279 System.err              pid-14087                            W  	at org.kiwix.kiwixmobile.core.StorageObserver.convertToBookOnDisk(StorageObserver.kt:57)
2024-05-03 12:11:37.367 14087-14279 System.err              pid-14087                            W  	at org.kiwix.kiwixmobile.core.StorageObserver.getBooksOnFileSystem$lambda-0(StorageObserver.kt:45)
2024-05-03 12:11:37.367 14087-14279 System.err              pid-14087                            W  	at org.kiwix.kiwixmobile.core.StorageObserver.$r8$lambda$J3Q4dPQVOFEOgCBh9ahA4Ca-Zak(Unknown Source:0)
2024-05-03 12:11:37.368 14087-14279 System.err              pid-14087                            W  	at org.kiwix.kiwixmobile.core.StorageObserver$$ExternalSyntheticLambda1.apply(Unknown Source:4)
2024-05-03 12:11:37.368 14087-14279 System.err              pid-14087                            W  	at io.reactivex.internal.operators.flowable.FlowableMap$MapConditionalSubscriber.onNext(FlowableMap.java:106)
2024-05-03 12:11:37.368 14087-14279 System.err              pid-14087                            W  	at io.reactivex.subscribers.SerializedSubscriber.onNext(SerializedSubscriber.java:100)
2024-05-03 12:11:37.368 14087-14279 System.err              pid-14087                            W  	at io.reactivex.internal.operators.flowable.FlowableWithLatestFrom$WithLatestFromSubscriber.tryOnNext(FlowableWithLatestFrom.java:94)
2024-05-03 12:11:37.368 14087-14279 System.err              pid-14087                            W  	at io.reactivex.internal.operators.flowable.FlowableWithLatestFrom$WithLatestFromSubscriber.onNext(FlowableWithLatestFrom.java:76)
2024-05-03 12:11:37.368 14087-14279 System.err              pid-14087                            W  	at io.reactivex.internal.operators.flowable.FlowableSubscribeOn$SubscribeOnSubscriber.onNext(FlowableSubscribeOn.java:97)
2024-05-03 12:11:37.368 14087-14279 System.err              pid-14087                            W  	at io.reactivex.internal.operators.flowable.FlowableCombineLatest$CombineLatestCoordinator.drainAsync(FlowableCombineLatest.java:374)
2024-05-03 12:11:37.369 14087-14279 System.err              pid-14087                            W  	at io.reactivex.internal.operators.flowable.FlowableCombineLatest$CombineLatestCoordinator.drain(FlowableCombineLatest.java:406)
2024-05-03 12:11:37.369 14087-14279 System.err              pid-14087                            W  	at io.reactivex.internal.operators.flowable.FlowableCombineLatest$CombineLatestCoordinator.innerValue(FlowableCombineLatest.java:250)
2024-05-03 12:11:37.369 14087-14279 System.err              pid-14087                            W  	at io.reactivex.internal.operators.flowable.FlowableCombineLatest$CombineLatestInnerSubscriber.onNext(FlowableCombineLatest.java:521)
2024-05-03 12:11:37.369 14087-14279 System.err              pid-14087                            W  	at io.reactivex.internal.operators.flowable.FlowableSubscribeOn$SubscribeOnSubscriber.onNext(FlowableSubscribeOn.java:97)
2024-05-03 12:11:37.369 14087-14279 System.err              pid-14087                            W  	at io.reactivex.internal.subscriptions.DeferredScalarSubscription.complete(DeferredScalarSubscription.java:132)
2024-05-03 12:11:37.370 14087-14279 System.err              pid-14087                            W  	at io.reactivex.internal.operators.flowable.FlowableFromCallable.subscribeActual(FlowableFromCallable.java:50)
2024-05-03 12:11:37.370 14087-14279 System.err              pid-14087                            W  	at io.reactivex.Flowable.subscribe(Flowable.java:14935)
2024-05-03 12:11:37.370 14087-14279 System.err              pid-14087                            W  	at io.reactivex.Flowable.subscribe(Flowable.java:14882)
2024-05-03 12:11:37.370 14087-14279 System.err              pid-14087                            W  	at io.reactivex.internal.operators.flowable.FlowableSubscribeOn$SubscribeOnSubscriber.run(FlowableSubscribeOn.java:82)
2024-05-03 12:11:37.370 14087-14279 System.err              pid-14087                            W  	at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
2024-05-03 12:11:37.370 14087-14279 System.err              pid-14087                            W  	at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
2024-05-03 12:11:37.371 14087-14279 System.err              pid-14087                            W  	at java.util.concurrent.FutureTask.run(FutureTask.java:264)
2024-05-03 12:11:37.371 14087-14279 System.err              pid-14087                            W  	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:307)
```

**Comment by MohitMaliFtechiz** [comment]
I just saw that this issue with zim-tool has now been fixed https://github.com/openzim/zim-tools/issues/401

**Comment by kelson42** [comment]
@MohitMaliFtechiz The bug of creating surnumerous empty chunk has been fixed, see https://github.com/openzim/zim-tools/issues/401. But, the fact that the libzim does not deliver properly if there is such an empty chunk is a bug. Please open an issue for that at openzim/libzim.

Regarding the workaround, you have developed to handle this libzim bug. Please keep it and put clear comment it should be removed once the libzim bug fixed. Open a dedicated issue to do that in next release.

... and the put this PR in review so we can mergw.

**Comment by MohitMaliFtechiz** [comment]
> @MohitMaliFtechiz The bug of creating surnumerous empty chunk has been fixed, see https://github.com/openzim/zim-tools/issues/401. But, the fact that the libzim does not deliver properly if there is such an empty chunk is a bug. Please open an issue for that at openzim/libzim.

@kelson42 Okay.

> Regarding the workaround, you have developed to handle this libzim bug. Please keep it and put clear comment it should be removed once the libzim bug fixed. Open a dedicated issue to do that in next release.

I have just captured the error thrown by the libzim to not crash the rendering process https://github.com/kiwix/kiwix-android/pull/3803/commits/ac7e0f662b0234eb5ba84ca134b87d843b9ac4a1 there is no other code changes I have done, okay I am putting the comments for this.

> ... and the put this PR in review so we can merge.

After integrating the latest `java-libkiwix` this PR is ready for merging. I will mark the PR ready then.

**Comment by mgautierfr** [comment]
Agree with @kelson42, libzim should be able to read split file even with a empty part at the end (or in the middle). I personally can open such split file without any issue (not using kiwix-android)

>  I have encountered an error in Android with zim-tools' splitted zim file, it was throwing the `Corrupted block detected` error when getting the 

`Corrupted block deteted` is not a libzim error. We have to investigate this but it may be not a libzim issue.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/reader/ZimFileReaderWithSplittedZimFileTest.kt [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+            File(storageDir, "testzim.zima${('a' + partNumber - 1).toChar()}")
+          )
+          partNumber++
+          currentPartSize = 0
+        } else {
+          outputStream = FileOutputStream(
+            File(storageDir, "testzim.zima${('a' + partNumber - 2).toChar()}"),
+            true
+          )
+        }
```
[the remark this answers is not in the document]
Comment (mgautierfr) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/reader/ZimFileReaderWithSplittedZimFileTest.kt:135]: And we may have corruption here too.

Content written through FileOutputStream may no be flushed right away. We have to call `close` (or `flush`) to be sure that all data is actually written.

By creating a new stream on the same file, without explicitly closing it means that during a small time, we have two FileOutputStream opened and writing on the same file (and mixing the content).

**Comment by MohitMaliFtechiz** [comment]
> Agree with @kelson42, libzim should be able to read split file even with a empty part at the end (or in the middle). I personally can open such split file without any issue (not using kiwix-android)
I have encountered an error in Android with zim-tools' splitted zim file, it was throwing the Corrupted block detected error when getting the
Corrupted block deteted is not a libzim error. We have to investigate this but it may be not a libzim issue.

@mgautierfr The error was throwing when I was trying to get the [mediaCount](https://github.com/kiwix/kiwix-android/blob/6656d941156bfc3e1e105e524f452a481655f36f/core/src/main/java/org/kiwix/kiwixmobile/core/reader/ZimFileReader.kt#L162) and [articleCount](https://github.com/kiwix/kiwix-android/blob/6656d941156bfc3e1e105e524f452a481655f36f/core/src/main/java/org/kiwix/kiwixmobile/core/reader/ZimFileReader.kt#L169) which are the `Archive's` functions.

> Corrupted block deteted is not a libzim error. We have to investigate this but it may be not a libzim issue.

If it is not a libzim error, and you can open the zim file of this type. Then it is probably the error of `java-libkiwix` binding since it is the middleware of libzim and `kiwix-android`. Have you ever tested this scenario with `java-libkiwix`?

**Comment by mgautierfr** [comment]
> `Corrupted block detected` is not a libzim error. We have to investigate this but it may be not a libzim issue.

This may be a zstd error (https://github.com/facebook/zstd/issues/2852). If it is the case, it means that we are passing corrupted data to be decompressed. It would be surprising that libzim pass a wrong value (assuming file is actually not corrupted), especially that a zim split made with `zimsplit` doesn't cut in the middle of a cluster.

However, your test split zim file creator is (potentially) splitting in the middle of cluster and I have identified a race condition. Are you sure the error is with a extra empty part or it is with the testing split file ?

File: app/src/androidTest/java/org/kiwix/kiwixmobile/reader/ZimFileReaderWithSplittedZimFileTest.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (MohitMaliFtechiz) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/reader/ZimFileReaderWithSplittedZimFileTest.kt]: @mgautierfr Thanks i have made the changes in test case.

**Comment by mgautierfr** [comment]
Ok, let's me try to resume the issue:

- In all case, you open zim file passing the path `wikipedia_ar_all_nopic_2023-11.zim`
- Split zim file with extra 0 sized part:
	. libzim 9.2.0, you have a crash
    . libzim 9.2.1, you don't have a crash
- Split zim file without extra 0 sized part:
    . It works with both version of libzim.

Are we agree on this ?

**Comment by MohitMaliFtechiz** [comment]
@mgautierfr Let me correct few things then you can resume the issue.

> In all case, you open zim file passing the path wikipedia_ar_all_nopic_2023-11.zim

I am opening the `wikipedia_ar_all_nopic_2023-11.zimaa` not `wikipedia_ar_all_nopic_2023-11.zim`.

> Split zim file with extra 0 sized part:
. libzim 9.2.0, you have a crash
. libzim 9.2.1, you don't have a crash

Libzim `9.2.0` does not opens in both situations whether there is extra 0 size zim file present or not.
With Libzim `9.2.1` there is a crash(when getting the value from these two methods `getArticleCount()` and `getMediaCount()`) when there is extra 0 size zim file present in storage, and without this 0 size zim file there is no crash. 

> Split zim file without extra 0 sized part:
. It works with both version of libzim.

No , in version 9.2.1 it working and in version 9.2.0 it is opening the zim file.

**Comment by kelson42** [comment]
@MohitMaliFtechiz I have released java-libkiwix 2.1.1, configured in kiwix-android and rebase this PR… but the CI is still red!

**Comment by mgautierfr** [comment]
Thank @MohitMaliFtechiz for the explanation.
How the split files are generated ? With zimsplit ? (Which version ?).
Can you run a `sha256sum` (or any other checksum tool) on them ?

**Comment by MohitMaliFtechiz** [comment]
> Thank @MohitMaliFtechiz for the explanation.
How the split files are generated ? With zimsplit ? (Which version ?).

I am using `zim-tools 3.1.0` for splitting the zim file.

```kotlin
zim-tools 3.1.0

libzim 7.2.0
+ libzstd 1.4.8
+ liblzma 5.2.5
+ libxapian 1.4.18
+ libicu 70.1.0
```

**Comment by kelson42** [comment]
@MohitMaliFtechiz Please redo the operation wirh latest version of zim-tools, your version is already a bit old

**Comment by MohitMaliFtechiz** [comment]
> Can you run a sha256sum (or any other checksum tool) on them ?

@mgautierfr Here are the all details from beginning.

* First I have tried to split the zim file. I have run the `zimsplit wikipedia_ar_all_nopic_2024-03.zim` command.
The output is:

```kotlin
acerpc@acerpc-Aspire-Lite-AL15-51:~/Documents$ zimsplit wikipedia_ar_all_nopic_2024-03.zim 
opening new file wikipedia_ar_all_nopic_2024-03.zimaa
opening new file wikipedia_ar_all_nopic_2024-03.zimab
opening new file wikipedia_ar_all_nopic_2024-03.zimac
opening new file wikipedia_ar_all_nopic_2024-03.zimad
WARNING: Part wikipedia_ar_all_nopic_2024-03.zimad is bigger that max part size. (3187373325>2147483648)
opening new file wikipedia_ar_all_nopic_2024-03.zimae
```

* Checksum report:
**Sha256sum**
```kotlin
acerpc@acerpc-Aspire-Lite-AL15-51:~/Documents$ sha256sum wikipedia_ar_all_nopic_2024-03.zim
f744022ed8208a50a1ce191cd5af946debc4012d36c41f7c338e8ae904b86303  wikipedia_ar_all_nopic_2024-03.zim
acerpc@acerpc-Aspire-Lite-AL15-51:~/Documents$ sha256sum wikipedia_ar_all_nopic_2024-03.zimaa
8fd91f8f328ca1a5c77129fd7e4d7435f663fa632e9c0078c3d38d0bccdea366  wikipedia_ar_all_nopic_2024-03.zimaa
acerpc@acerpc-Aspire-Lite-AL15-51:~/Documents$ sha256sum wikipedia_ar_all_nopic_2024-03.zimab
eb1892b3bfee9876626a12bff57d90c6663ab143e74d88e8ebddc6084f167e82  wikipedia_ar_all_nopic_2024-03.zimab
acerpc@acerpc-Aspire-Lite-AL15-51:~/Documents$ sha256sum wikipedia_ar_all_nopic_2024-03.zimac
33729314abe6b640d918770cfd9146a13b49eac537a52b00e6568e3a358d1ce5  wikipedia_ar_all_nopic_2024-03.zimac
acerpc@acerpc-Aspire-Lite-AL15-51:~/Documents$ sha256sum wikipedia_ar_all_nopic_2024-03.zimad
b17e5f6fbbc140486cab4c19adc8970364796acdbd06a8a9439e501d02cfb040  wikipedia_ar_all_nopic_2024-03.zimad
acerpc@acerpc-Aspire-Lite-AL15-51:~/Documents$ sha256sum wikipedia_ar_all_nopic_2024-03.zimae
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855  wikipedia_ar_all_nopic_2024-03.zimae
acerpc@acerpc-Aspire-Lite-AL15-51:~/Documents$ 
```

**md5sum**

```kotlin
acerpc@acerpc-Aspire-Lite-AL15-51:~/Documents$ md5sum wikipedia_ar_all_nopic_2024-03.zim
66bb7f6ca0008b304231040a53b71ab4  wikipedia_ar_all_nopic_2024-03.zim
acerpc@acerpc-Aspire-Lite-AL15-51:~/Documents$ md5sum wikipedia_ar_all_nopic_2024-03.zimaa
981bff8a335e0e13ac2a9fd1af25df43  wikipedia_ar_all_nopic_2024-03.zimaa
acerpc@acerpc-Aspire-Lite-AL15-51:~/Documents$ md5sum wikipedia_ar_all_nopic_2024-03.zimab
4ff359e0dbdd1af19ce39353f198e31d  wikipedia_ar_all_nopic_2024-03.zimab
acerpc@acerpc-Aspire-Lite-AL15-51:~/Documents$ md5sum wikipedia_ar_all_nopic_2024-03.zimac
6504996d96eacc34e9c504c1ca4d0acb  wikipedia_ar_all_nopic_2024-03.zimac
acerpc@acerpc-Aspire-Lite-AL15-51:~/Documents$ md5sum wikipedia_ar_all_nopic_2024-03.zimad
a909742cd7df2e8aa7dffe9f9d74dd3f  wikipedia_ar_all_nopic_2024-03.zimad
acerpc@acerpc-Aspire-Lite-AL15-51:~/Documents$ md5sum wikipedia_ar_all_nopic_2024-03.zimae
d41d8cd98f00b204e9800998ecf8427e  wikipedia_ar_all_nopic_2024-03.zimae
```

**cksum**
```kotlin
acerpc@acerpc-Aspire-Lite-AL15-51:~/Documents$ cksum wikipedia_ar_all_nopic_2024-03.zim
3675923895 9262630633 wikipedia_ar_all_nopic_2024-03.zim
acerpc@acerpc-Aspire-Lite-AL15-51:~/Documents$ cksum wikipedia_ar_all_nopic_2024-03.zimaa
1205357161 2147459163 wikipedia_ar_all_nopic_2024-03.zimaa
acerpc@acerpc-Aspire-Lite-AL15-51:~/Documents$ cksum wikipedia_ar_all_nopic_2024-03.zimab
3126437707 2147376980 wikipedia_ar_all_nopic_2024-03.zimab
acerpc@acerpc-Aspire-Lite-AL15-51:~/Documents$ cksum wikipedia_ar_all_nopic_2024-03.zimac
3803321215 1780421165 wikipedia_ar_all_nopic_2024-03.zimac
acerpc@acerpc-Aspire-Lite-AL15-51:~/Documents$ cksum wikipedia_ar_all_nopic_2024-03.zimad
2606185103 3187373325 wikipedia_ar_all_nopic_2024-03.zimad
acerpc@acerpc-Aspire-Lite-AL15-51:~/Documents$ cksum wikipedia_ar_all_nopic_2024-03.zimae
4294967295 0 wikipedia_ar_all_nopic_2024-03.zimae
```

* One more thing the last chunk of the zim file is `text document` type, other zim chunks are `Binary` type.

1. wikipedia_ar_all_nopic_2023-11.zimaa (2.1 GB (2,14,74,74,503 bytes) **Type** = Binary (application/octet-stream)
2. wikipedia_ar_all_nopic_2023-11.zimab (2.1 GB (2,14,74,71,095 bytes) **Type** = Binary (application/octet-stream)
3. wikipedia_ar_all_nopic_2023-11.zimac (1.3 GB (1,28,14,26,182 bytes) **Type** = Binary (application/octet-stream)
4. wikipedia_ar_all_nopic_2023-11.zimad (3.1 GB (3,14,05,16,010 bytes) **Type** = Binary (application/octet-stream)
5. wikipedia_ar_all_nopic_2023-11.zimae (0 bytes) **Type** = plain text document (text/plain)

Screenshot from 2024-05-06 12-21-13

**Comment by MohitMaliFtechiz** [comment]
> @MohitMaliFtechiz Please redo the operation wirh latest version of zim-tools, your version is already a bit old... maybe there a bug arounf this

@kelson42 The latest version of `zim-tools` is available for my current version of Ubuntu(Ubuntu 22.04.4 LTS)  is [3.1.0-3](https://manpages.ubuntu.com/manpages/jammy/man1/zimsplit.1.html) which I am using.

The latest version of `zim-tools` requires Ubuntu `23.10` or `24.04` https://manpages.ubuntu.com/manpages/mantic/man1/zimsplit.1.html. Is it any other way to use the latest `zim-tools` without upgrading the ubuntu.

**Comment by mgautierfr** [comment]
You can get the last version of zim-tools here : https://download.openzim.org/release/zim-tools/

But anyway, the checksum of you file are same as mine, so it is not a problem of zim-tools version.

**Comment by mgautierfr** [comment]
You can get the last version of zim-tools here : https://download.openzim.org/release/zim-tools/

But anyway, the checksum of you file are same as mine, so it is not a problem of zim-tools version.

I have add a simple (java) test opening split files (with an extra 0 part) on my computer and it works. So it seems it is a android only issue.

**Comment by mgautierfr** [comment]
Can you try to run the application under strace (https://source.android.com/docs/core/tests/debug/strace) please ?

**Comment by MohitMaliFtechiz** [comment]
> You can get the last version of zim-tools here : https://download.openzim.org/release/zim-tools/
But anyway, the checksum of you file are same as mine, so it is not a problem of zim-tools version.
I have add a simple (java) test opening split files (with an extra 0 part) on my computer and it works. So it seems it is a android only issue.

@mgautierfr Thanks.

**Comment by MohitMaliFtechiz** [comment]
@mgautierfr Thanks for your testing, I have also added the test case in Android with 0 bytes zim file, and my test case is passed without any error as well, it is returning the `articleCount` and `mediaCount` both in my test case. However, the error is with the `zim-tools` splitted zim file remains the same.

**Comment by kelson42** [comment]
@MohitMaliFtechiz The handling of this edge case is dlowing down the release. Please create dedicated issue to continue the work, so we can merge this one.

**Comment by MohitMaliFtechiz** [comment]
> @MohitMaliFtechiz The handling of this edge case is dlowing down the release. Please create dedicated issue to continue the work, so we can merge this one.

@kelson42 I have created a separate ticket for this edge case https://github.com/kiwix/kiwix-android/issues/3827.

**Comment by kelson42** [comment]
@MohitMaliFtechiz Please don't call a PR by the error... but by the solution! Here this PR has an issue title!
